### PR TITLE
add NativeCryptoDispensary.sol for magic links which drop the native currency

### PR DESCRIPTION
**Pull request by James-Sangalli**
This PR creates a contract which enables for magic link drops of native currency like Ether or xDAI. 

Edit: aa72834 working on ropsten

@zhangzhongnan928 ready to move forward with DAI technical team, have this sample contract, payment master sample (either the old one or the new JS one, depends on requirements from them), iOS sample working: https://github.com/AlphaWallet/alpha-wallet-ios/pull/1053 and an amendment to Peter's tool which generates magic links for this: https://github.com/lyhistory/Peter-s-Tool/pull/18

### CODE REVIEW

File: spawnable-contract/currency-links/NativeCryptoDispensary.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
+//magic link encoding 0x04
+//send any native currency to this contract and use the contract balance to move the funds
+//Ropsten: 0x9d928a678eeaaEfA19eF73E5368830b3476C0678
+pragma solidity 0.5.1;
+contract NativeCryptoDispensary {
+
+    bytes8 public requiredPrefix = "XDAIDROP";
+    address payable approvedSigner;
+    uint[] nonces;
```
Comment (hboon) [inline spawnable-contract/currency-links/NativeCryptoDispensary.sol]: Maybe?
```
uint32[] nonces;
```

File: spawnable-contract/currency-links/NativeCryptoDispensary.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    ) public {
+        //stop people from stealing drops by enforcing a paymaster to use it
+        require(msg.sender == approvedPaymaster);
+        require(block.timestamp < expiry);
+        checkIfNonceUsed(nonce);
+        //message is signed with szabo price 
+        bytes32 message = formMessage(nonce, amount, expiry, address(this));
+        require(ecrecover(message, v, r, s) == approvedSigner);
+        //price is signed as szabo units 
+        receiver.transfer(amount * (1 szabo));
```
Comment (hboon) [inline spawnable-contract/currency-links/NativeCryptoDispensary.sol]: Should we add the `nonce` to `nonces` before/after the transfer?

File: spawnable-contract/currency-links/NativeCryptoDispensary.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        }
+        for(uint i = 0; i < 20; i++) {
+            message[i + 20] = byte(bytes20(bytes20(contractAddress) << (8 * i)));
+        }
+        return keccak256(message);
+    }
+    
+    function checkIfNonceUsed(uint nonce) internal view {
+        for(uint i = 0; i < nonces.length; i++) {
+            if(nonces[i] == nonce) {
```
Comment (hboon) [inline spawnable-contract/currency-links/NativeCryptoDispensary.sol]: O(n) operation here.

File: spawnable-contract/currency-links/NativeCryptoDispensary.sol [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            message[i + 8] = byte(bytes4(nonce << (8 * i)));
+        }
+        for(uint i = 0; i < 4; i++) {
+            message[i + 12] = byte(bytes4(amount << (8 * i)));
+        }
+        for(uint i = 0; i < 4; i++) {
+            message[i + 16] = byte(bytes4(expiry << (8 * i)));
+        }
+        for(uint i = 0; i < 20; i++) {
+            message[i + 20] = byte(bytes20(bytes20(contractAddress) << (8 * i)));
```
Comment (hboon) [inline spawnable-contract/currency-links/NativeCryptoDispensary.sol]: Only need to do `bytes20()` once.

Comment (James-Sangalli) [reply spawnable-contract/currency-links/NativeCryptoDispensary.sol]: @hboon new compiler requires this conversion, can no longer do it directly on an address variable

Comment (hboon) [reply spawnable-contract/currency-links/NativeCryptoDispensary.sol]: This should work?

```
message[i + 20] = byte(bytes20(contractAddress) << (8 * i));
```

**Pull request merged**